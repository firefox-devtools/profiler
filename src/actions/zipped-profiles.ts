/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
import {
  getZipFileTable,
  getZipFileState,
} from 'firefox-profiler/selectors/zipped-profiles';
import { unserializeProfileOfArbitraryFormat } from 'firefox-profiler/profile-logic/process-profile';
import { loadProfile } from './receive-profile';

import { Action, ThunkAction } from 'firefox-profiler/types';
import { IndexIntoZipFileTable } from 'firefox-profiler/profile-logic/zip-files';

export function changeSelectedZipFile(
  selectedZipFileIndex: IndexIntoZipFileTable
): Action {
  return {
    type: 'CHANGE_SELECTED_ZIP_FILE' as const,
    selectedZipFileIndex,
  };
}

export function changeExpandedZipFile(
  expandedZipFileIndexes: Array<IndexIntoZipFileTable | null>
): Action {
  return {
    type: 'CHANGE_EXPANDED_ZIP_FILES' as const,
    expandedZipFileIndexes,
  };
}

/**
 * This ThunkAction has a bit of complexity to it, due to the asynchronous nature of
 * reading in profiles from a zip file. The UrlState represents the current desired state
 * of what file to view in a zip file, and the ZipFileState represents the actual steps
 * being performed to reach the desired UrlState.
 *
 * This ThunkAction needs to properly handle when the UrlState changes faster than it
 * can change the ZipFileState, and not have any race conditions.
 */
export function viewProfileFromZip(
  zipFileIndex: IndexIntoZipFileTable,
  initialLoad: boolean = false
): ThunkAction<Promise<void>> {
  return async (dispatch, getState) => {
    const zipFileTable = getZipFileTable(getState());
    const pathInZipFile = zipFileTable.path[zipFileIndex];
    const file = zipFileTable.file[zipFileIndex];
    if (!file) {
      throw new Error(
        'Attempted to load an archive that did not exist or was a directory.'
      );
    }

    dispatch({ type: 'PROCESS_PROFILE_FROM_ZIP_FILE' as const, pathInZipFile });

    try {
      // Attempt to unserialize the profile.
      const profile = await unserializeProfileOfArbitraryFormat(
        await file.async('string'),
        pathInZipFile
      );

      // Since this is an async function, there can be race conditions. Prevent this by
      // comparing this request with the current state of the store. If this result
      // is invalid, don't dispatch anything, and discard the profile.
      const zipFileState = getZipFileState(getState());
      if (
        zipFileState.pathInZipFile === pathInZipFile &&
        zipFileState.phase === 'PROCESS_PROFILE_FROM_ZIP_FILE'
      ) {
        await dispatch(loadProfile(profile, { pathInZipFile }, initialLoad));
      }
    } catch (error) {
      console.error(
        'Failed to process the profile in the archive with the following error:'
      );
      console.error(error);
      dispatch({
        type: 'FAILED_TO_PROCESS_PROFILE_FROM_ZIP_FILE' as const,
        error,
      });
    }
  };
}

/**
 * This function can take a zip file path, but the path can come from the URL, so
 * don't really trust it.
 */
export function viewProfileFromPathInZipFile(
  pathInZipFile: string
): ThunkAction<Promise<void>> {
  return (dispatch, getState) => {
    const zipFileTable = getZipFileTable(getState());
    const zipFileIndex = zipFileTable.path.indexOf(pathInZipFile);
    if (zipFileIndex === -1) {
      dispatch(showErrorForNoFileInZip(pathInZipFile));
      return Promise.resolve();
    }
    return dispatch(viewProfileFromZip(zipFileIndex));
  };
}

export function returnToZipFileList() {
  return { type: 'RETURN_TO_ZIP_FILE_LIST' as const };
}

export function showErrorForNoFileInZip(pathInZipFile: string) {
  return { type: 'FILE_NOT_FOUND_IN_ZIP_FILE' as const, pathInZipFile };
}
