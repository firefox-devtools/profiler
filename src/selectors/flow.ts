/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { createSelector } from 'reselect';

import {
  computeProfileFlowInfo,
  computeFlowTiming,
} from '../profile-logic/marker-data';
import { getThreadSelectors } from './per-thread';
import { getActiveFlows } from './url-state';
import type { ThreadSelectors } from './per-thread';
import {
  getThreads,
  getMarkerSchema,
  getRawProfileSharedData,
} from './profile';

import type {
  Selector,
  State,
  MarkerIndex,
  Marker,
  ProfileFlowInfo,
  FlowTiming,
} from 'firefox-profiler/types';

function _arraysShallowEqual(arr1: any[], arr2: any[]): boolean {
  return arr1.length === arr2.length && arr1.every((val, i) => val === arr2[i]);
}

function _createSelectorForAllThreads<T>(
  f: (selectors: ThreadSelectors, state: State) => T
): Selector<T[]> {
  let previousOutputPerThread: T[] = [];
  return function recomputeSelectorForAllThreads(state: State): T[] {
    const threads = getThreads(state);
    let outputPerThread = threads.map((_thread, i) => {
      const threadSelectors = getThreadSelectors(i);
      return f(threadSelectors, state);
    });
    if (_arraysShallowEqual(outputPerThread, previousOutputPerThread)) {
      outputPerThread = previousOutputPerThread;
    }
    previousOutputPerThread = outputPerThread;
    return outputPerThread;
  };
}

export const getFullMarkerListPerThread: Selector<Marker[][]> =
  _createSelectorForAllThreads(({ getFullMarkerList }, state) =>
    getFullMarkerList(state)
  );

export const getMarkerChartLabelGetterPerThread: Selector<
  Array<(marker: MarkerIndex) => string>
> = _createSelectorForAllThreads(({ getMarkerChartLabelGetter }, state) =>
  getMarkerChartLabelGetter(state)
);

export const getProfileFlowInfo: Selector<ProfileFlowInfo> = createSelector(
  getFullMarkerListPerThread,
  getThreads,
  getMarkerSchema,
  getRawProfileSharedData,
  computeProfileFlowInfo
);

export const getFlowTiming: Selector<FlowTiming> = createSelector(
  getProfileFlowInfo,
  getActiveFlows,
  computeFlowTiming
);
