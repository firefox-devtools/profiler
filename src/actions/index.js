/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as app from './app';
import * as icons from './icons';
import * as profileView from './profile-view';
import * as publish from './publish';
import * as receiveProfile from './receive-profile';
import * as zippedProfiles from './zipped-profiles';
import * as code from './code';

export default Object.assign(
  {},
  app,
  icons,
  profileView,
  publish,
  receiveProfile,
  zippedProfiles,
  code
);
