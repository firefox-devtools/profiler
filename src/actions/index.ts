/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
import * as app from './app';
import * as icons from './icons';
import * as l10n from './l10n';
import * as profileView from './profile-view';
import * as publish from './publish';
import * as receiveProfile from './receive-profile';
import * as zippedProfiles from './zipped-profiles';
import * as code from './code';

export default Object.assign(
  {},
  app,
  icons,
  l10n,
  profileView,
  publish,
  receiveProfile,
  zippedProfiles,
  code
);
