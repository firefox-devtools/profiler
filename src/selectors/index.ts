/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
export * from './app';
export * from './l10n';
export * from './per-thread';
export * from './profile';
export * from './url-state';
export * from './icons';
export * from './publish';
export * from './zipped-profiles';
export * from './cpu';
export * from './code';

import * as app from './app';
import {
  selectedThreadSelectors as selectedThread,
  selectedNodeSelectors as selectedNode,
} from './per-thread';
import * as profile from './profile';
import * as urlState from './url-state';
import * as icons from './icons';
import * as publish from './publish';
import * as zippedProfiles from './zipped-profiles';
import * as l10n from './l10n';
import * as cpu from './cpu';
import * as code from './code';

const _selectorsForConsole = {
  app,
  profile,
  urlState,
  icons,
  publish,
  zippedProfiles,
  selectedThread,
  selectedNode,
  l10n,
  cpu,
  code,
};

// Exports require explicit typing. Deduce the type with typeof.
export const selectorsForConsole: typeof _selectorsForConsole =
  _selectorsForConsole;
