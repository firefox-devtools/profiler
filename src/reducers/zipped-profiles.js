/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { combineReducers } from 'redux';
import { oneLine } from 'common-tags';
import { ensureExists } from '../utils/flow';
import * as ZipFiles from '../profile-logic/zip-files';

import type {
  ZipFileState,
  Reducer,
  ZippedProfilesState,
} from 'firefox-profiler/types';

/**
 * This reducer contains all of the state that deals with loading in profiles from
 * zip files.
 */
const zipFile: Reducer<ZipFileState> = (
  state = {
    phase: 'NO_ZIP_FILE',
    zip: null,
    pathInZipFile: null,
  },
  action
) => {
  switch (action.type) {
    case 'UPDATE_URL_STATE': {
      if (
        action.newUrlState &&
        action.newUrlState.pathInZipFile === null &&
        state.phase === 'VIEW_PROFILE_IN_ZIP_FILE'
      ) {
        // When the back button is hit on the browser, the UrlState can update, but
        // the state of the zip file viewer can be out of date. In this case, make
        // sure and revert the state back to LIST_FILES_IN_ZIP_FILE rather than
        // VIEW_PROFILE_IN_ZIP_FILE. Otherwise the zip file viewer will try and
        // display a profile that does not exist.
        return _validateStateTransition(state, {
          phase: 'LIST_FILES_IN_ZIP_FILE',
          zip: state.zip,
          pathInZipFile: null,
        });
      }
      return state;
    }
    case 'RECEIVE_ZIP_FILE':
      return _validateStateTransition(state, {
        phase: 'LIST_FILES_IN_ZIP_FILE',
        zip: action.zip,
        pathInZipFile: null,
      });
    case 'RETURN_TO_ZIP_FILE_LIST':
    case 'DISMISS_PROCESS_PROFILE_FROM_ZIP_ERROR':
      return _validateStateTransition(state, {
        phase: 'LIST_FILES_IN_ZIP_FILE',
        zip: ensureExists(state.zip),
        pathInZipFile: null,
      });
    case 'PROCESS_PROFILE_FROM_ZIP_FILE':
      return _validateStateTransition(state, {
        phase: 'PROCESS_PROFILE_FROM_ZIP_FILE',
        zip: ensureExists(state.zip),
        pathInZipFile: action.pathInZipFile,
      });
    case 'FILE_NOT_FOUND_IN_ZIP_FILE':
      return _validateStateTransition(state, {
        phase: 'FILE_NOT_FOUND_IN_ZIP_FILE',
        zip: ensureExists(state.zip),
        pathInZipFile: action.pathInZipFile,
      });
    case 'FAILED_TO_PROCESS_PROFILE_FROM_ZIP_FILE':
      return _validateStateTransition(state, {
        phase: 'FAILED_TO_PROCESS_PROFILE_FROM_ZIP_FILE',
        zip: ensureExists(state.zip),
        pathInZipFile: ensureExists(state.pathInZipFile),
      });
    case 'FATAL_ERROR':
      return state.phase === 'NO_ZIP_FILE' ||
        state.phase === 'LIST_FILES_IN_ZIP_FILE'
        ? state
        : _validateStateTransition(state, {
            phase: 'FAILED_TO_PROCESS_PROFILE_FROM_ZIP_FILE',
            zip: ensureExists(state.zip),
            pathInZipFile: ensureExists(state.pathInZipFile),
          });
    case 'VIEW_FULL_PROFILE':
      // Only process this as a change if a zip file is actually loaded.
      return state.phase === 'NO_ZIP_FILE'
        ? state
        : _validateStateTransition(state, {
            phase: 'VIEW_PROFILE_IN_ZIP_FILE',
            zip: ensureExists(state.zip),
            pathInZipFile: ensureExists(state.pathInZipFile),
          });
    case 'PROFILE_PUBLISHED':
    case 'SANITIZED_PROFILE_PUBLISHED':
    case 'WAITING_FOR_PROFILE_FROM_FILE':
      // Revert back to not having a zip file.
      return {
        phase: 'NO_ZIP_FILE',
        zip: null,
        pathInZipFile: null,
      };
    default:
      return state;
  }
};

const error: Reducer<null | Error> = (state = null, action) => {
  switch (action.type) {
    case 'FAILED_TO_PROCESS_PROFILE_FROM_ZIP_FILE':
    case 'FATAL_ERROR':
      return action.error;
    default:
      return state;
  }
};

/**
 * This function ensures that the state transitions are logical and make sense. The
 * switch statement provides a mapping of what states are valid to transition to from
 * a previous state. A switch is used rather than an object in order to make it easier
 * to exhaustively check with Flow.
 */
function _validateStateTransition(
  prev: ZipFileState,
  next: ZipFileState
): ZipFileState {
  const prevPhase = prev.phase;
  if (prevPhase === next.phase) {
    return next;
  }
  let expectedNextPhases;
  switch (prevPhase) {
    case 'NO_ZIP_FILE':
      expectedNextPhases = [
        // Coming into a fresh page load with a zip file, this is the first state
        // transition.
        'LIST_FILES_IN_ZIP_FILE',
        // This will happen if coming in from a URL that is already pointing at a file
        // in a zip.
        'PROCESS_PROFILE_FROM_ZIP_FILE',
      ];
      break;
    case 'LIST_FILES_IN_ZIP_FILE':
      expectedNextPhases = [
        'PROCESS_PROFILE_FROM_ZIP_FILE',
        'FILE_NOT_FOUND_IN_ZIP_FILE',
      ];
      break;
    case 'PROCESS_PROFILE_FROM_ZIP_FILE':
      expectedNextPhases = [
        'VIEW_PROFILE_IN_ZIP_FILE',
        'FAILED_TO_PROCESS_PROFILE_FROM_ZIP_FILE',
        // When navigating with the URL, it's possible to go back and list the files.
        'LIST_FILES_IN_ZIP_FILE',
      ];
      break;
    case 'FILE_NOT_FOUND_IN_ZIP_FILE':
      expectedNextPhases = ['LIST_FILES_IN_ZIP_FILE'];
      break;
    case 'FAILED_TO_PROCESS_PROFILE_FROM_ZIP_FILE':
      expectedNextPhases = ['LIST_FILES_IN_ZIP_FILE'];
      break;
    case 'VIEW_PROFILE_IN_ZIP_FILE':
      expectedNextPhases = ['LIST_FILES_IN_ZIP_FILE'];
      break;
    default:
      throw new Error(`Unhandled ZipFileState “${(prevPhase: empty)}”`);
  }
  if (!expectedNextPhases.includes(next.phase)) {
    console.error('Previous ZipFileState:', prev);
    console.error('Next ZipFileState:', next);

    throw new Error(oneLine`
      Attempted to transition the ZipFileState from the phase “${prev.phase}”
      to “${next.phase}”, however “${prev.phase}” can only transition to
      “${expectedNextPhases.join('”, “')}”.
    `);
  }
  return next;
}

const selectedZipFileIndex: Reducer<null | ZipFiles.IndexIntoZipFileTable> = (
  state = null,
  action
) => {
  switch (action.type) {
    case 'CHANGE_SELECTED_ZIP_FILE': {
      return action.selectedZipFileIndex;
    }
    default:
      return state;
  }
};

const expandedZipFileIndexes: Reducer<
  Array<ZipFiles.IndexIntoZipFileTable | null>,
> = (
  // In practice this should never contain null, but needs to support the
  // TreeView interface.
  state = [],
  action
) => {
  switch (action.type) {
    case 'CHANGE_EXPANDED_ZIP_FILES': {
      return action.expandedZipFileIndexes;
    }
    default:
      return state;
  }
};

const zipFileReducer: Reducer<ZippedProfilesState> = combineReducers({
  zipFile,
  error,
  selectedZipFileIndex,
  expandedZipFileIndexes,
});

export default zipFileReducer;
