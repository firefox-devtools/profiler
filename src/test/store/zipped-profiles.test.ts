/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
import {
  formatZipFileTable,
  storeWithZipFile,
} from '../fixtures/profiles/zip-file';
import * as ProfileViewSelectors from '../../selectors/profile';
import * as ZippedProfilesSelectors from '../../selectors/zipped-profiles';
import * as UrlStateSelectors from '../../selectors/url-state';
import createStore from '../../app-logic/create-store';
import { ensureExists } from '../../utils/flow';
import JSZip from 'jszip';

import * as ZippedProfilesActions from '../../actions/zipped-profiles';
import * as ReceiveProfileActions from '../../actions/receive-profile';
import * as ProfileViewActions from '../../actions/profile-view';
import type { PreviewSelection } from 'firefox-profiler/types';

describe('reducer zipFileState', function () {
  it('can store the zip file in the reducer', async function () {
    const { zippedProfiles } = await storeWithZipFile();
    expect(zippedProfiles).toBe(zippedProfiles);
  });

  it('can load a profile from the zip file', async function () {
    const { dispatch, getState } = await storeWithZipFile([
      'foo/bar/profile1.json',
      'foo/profile2.json',
      'baz/profile3.json',
    ]);
    expect(ProfileViewSelectors.getProfileOrNull(getState())).toEqual(null);

    await dispatch(
      ZippedProfilesActions.viewProfileFromPathInZipFile(
        'foo/bar/profile1.json'
      )
    );

    expect(ZippedProfilesSelectors.getZipFileState(getState()).phase).toBe(
      'VIEW_PROFILE_IN_ZIP_FILE'
    );
    const profile1 = ProfileViewSelectors.getProfile(getState());

    expect(profile1).toBeTruthy();
  });

  describe('profileName handling', function () {
    async function setup() {
      const { dispatch, getState } = await storeWithZipFile([
        'foo/bar/profile1.json',
        'foo/profile2.json',
        'baz/profile3.json',
        'profile4.json',
      ]);

      await dispatch(
        ZippedProfilesActions.viewProfileFromPathInZipFile(
          'foo/bar/profile1.json'
        )
      );

      return { getState, dispatch };
    }

    it('computes a profile name from the loaded file', async function () {
      const { getState } = await setup();

      const expectedName = 'bar/profile1.json';
      expect(UrlStateSelectors.getProfileNameWithDefault(getState())).toBe(
        expectedName
      );
    });

    it('computes a profile when no folder present', async function () {
      const { dispatch, getState } = await storeWithZipFile([
        'foo/bar/profile1.json',
        'foo/profile2.json',
        'baz/profile3.json',
        'profile4.json',
      ]);

      await dispatch(
        ZippedProfilesActions.viewProfileFromPathInZipFile('profile4.json')
      );

      const expectedName = 'profile4.json';
      expect(UrlStateSelectors.getProfileNameWithDefault(getState())).toBe(
        expectedName
      );
    });

    it('prefers ProfileName if it is given in the URL', async function () {
      const { getState, dispatch } = await setup();
      const profileNameFromURL = 'good profile';

      dispatch(ProfileViewActions.changeProfileName(profileNameFromURL));
      expect(UrlStateSelectors.getProfileNameWithDefault(getState())).toBe(
        profileNameFromURL
      );
    });
  });

  it('will fail when trying to load an invalid profile', async function () {
    const store = createStore();
    const { getState, dispatch } = store;
    const zip = new JSZip();
    zip.file('not-a-profile.json', 'not a profile');
    dispatch(ReceiveProfileActions.receiveZipFile(zip));

    jest.spyOn(console, 'error').mockImplementation(() => {});

    await dispatch(
      ZippedProfilesActions.viewProfileFromPathInZipFile('not-a-profile.json')
    );

    expect(ZippedProfilesSelectors.getZipFileState(getState()).phase).toEqual(
      'FAILED_TO_PROCESS_PROFILE_FROM_ZIP_FILE'
    );
    // console error was called.
    // @ts-expect-error - 'mock does not exist on type'
    expect(console.error.mock.calls.length >= 1).toEqual(true);
    // @ts-expect-error - 'mock does not exist on type'
    expect(console.error.mock.calls).toMatchSnapshot();
  });

  it('will fail when not finding a profile', async function () {
    const store = createStore();
    const { getState, dispatch } = store;
    dispatch(ReceiveProfileActions.receiveZipFile(new JSZip()));
    dispatch(
      ZippedProfilesActions.viewProfileFromPathInZipFile('nothing-here.json')
    );

    expect(ZippedProfilesSelectors.getZipFileState(getState()).phase).toEqual(
      'FILE_NOT_FOUND_IN_ZIP_FILE'
    );
  });

  it('can compute a ZipFileTable', async function () {
    const { getState } = await storeWithZipFile([
      'foo/bar/profile1.json',
      'foo/profile2.json',
      'foo/profile3.json',
      'foo/profile4.json',
      'baz/profile5.json',
    ]);

    const zipFileTable = ZippedProfilesSelectors.getZipFileTable(getState());
    expect(formatZipFileTable(zipFileTable)).toEqual([
      'foo (dir)',
      '  bar (dir)',
      '    profile1.json (file)',
      '  profile2.json (file)',
      '  profile3.json (file)',
      '  profile4.json (file)',
      'baz (dir)',
      '  profile5.json (file)',
    ]);
  });

  it('computes the zip file max depth', async function () {
    const { getState } = await storeWithZipFile([
      'foo/bar/profile1.json',
      'foo/profile2.json',
      'foo/profile3.json',
      'foo/profile4.json',
      'baz/profile5.json',
    ]);
    expect(ZippedProfilesSelectors.getZipFileMaxDepth(getState())).toEqual(2);
  });
});

describe('selected and expanded zip files', function () {
  it('can expand selections in the zip file', function () {
    const { dispatch, getState } = createStore();

    expect(
      ZippedProfilesSelectors.getExpandedZipFileIndexes(getState())
    ).toEqual([]);

    // The indexes don't check that they are valid when you add them.
    dispatch(ZippedProfilesActions.changeExpandedZipFile([123, 456, 789]));
    expect(
      ZippedProfilesSelectors.getExpandedZipFileIndexes(getState())
    ).toEqual([123, 456, 789]);
  });

  it('can procure an interesting selection', async function () {
    const { getState, dispatch } = await storeWithZipFile([
      'a/profile1.json',
      'a/profile2.json',
      'b/profile3.json',
      'b/profile4.json',
      'c/profile5.json',
      'c/profile6.json',
      'd/profile7.json',
      'd/profile8.json',
    ]);

    const zipFileTree = ZippedProfilesSelectors.getZipFileTree(getState());
    const zipFileTable = ZippedProfilesSelectors.getZipFileTable(getState());
    dispatch(
      ZippedProfilesActions.changeExpandedZipFile([
        ...zipFileTree.getAllDescendants(null),
      ])
    );
    const expanded =
      ZippedProfilesSelectors.getExpandedZipFileIndexes(getState());
    const expandedNames = expanded.map(
      (index) => zipFileTable.path[ensureExists(index)]
    );
    expect(expandedNames).toEqual([
      'a',
      'a/profile1.json',
      'a/profile2.json',
      'b',
      'b/profile3.json',
      'b/profile4.json',
      'c',
      'c/profile5.json',
      'c/profile6.json',
      'd',
      'd/profile7.json',
      'd/profile8.json',
    ]);
  });

  it('can select a zip file', function () {
    const { dispatch, getState } = createStore();

    expect(ZippedProfilesSelectors.getSelectedZipFileIndex(getState())).toEqual(
      null
    );

    // The indexes don't check that they are valid when you add them.
    dispatch(ZippedProfilesActions.changeSelectedZipFile(123));

    expect(ZippedProfilesSelectors.getSelectedZipFileIndex(getState())).toEqual(
      123
    );
  });
});

describe('profile state invalidation when switching between profiles', function () {
  const getStoreViewingProfile = async () => {
    const { dispatch, getState } = await storeWithZipFile([
      'profile1.json',
      'profile2.json',
    ]);

    const viewProfile = (path: string) =>
      dispatch(ZippedProfilesActions.viewProfileFromPathInZipFile(path));

    return { dispatch, getState, viewProfile };
  };

  it('invalidates profile-specific url state', async function () {
    const { dispatch, getState, viewProfile } = await getStoreViewingProfile();
    viewProfile('profile1.json');

    // It starts out empty.
    expect(UrlStateSelectors.getAllCommittedRanges(getState())).toEqual([]);

    // Add a url-encoded bit of state.
    dispatch(ProfileViewActions.commitRange(0, 10));
    expect(UrlStateSelectors.getAllCommittedRanges(getState())).toEqual([
      { start: 0, end: 10 },
    ]);

    // It switches to another profile and invalidates.
    dispatch(ZippedProfilesActions.returnToZipFileList());
    viewProfile('profile2.json');
    expect(UrlStateSelectors.getAllCommittedRanges(getState())).toEqual([]);
  });

  it('invalidates profile view state', async function () {
    const { dispatch, getState, viewProfile } = await getStoreViewingProfile();
    viewProfile('profile1.json');

    // Create new copies of the selection on each assertion and change, so
    // that we are not relying on strict equality.
    const getNoSelection = () => ({ hasSelection: false, isModifying: false });
    const getSomeSelection = () =>
      ({
        hasSelection: true,
        isModifying: true,
        selectionStart: 0,
        selectionEnd: 10,
      }) as PreviewSelection;

    // It starts with no selection.
    expect(ProfileViewSelectors.getPreviewSelection(getState())).toEqual(
      getNoSelection()
    );

    // Add a selection.
    dispatch(ProfileViewActions.updatePreviewSelection(getSomeSelection()));
    expect(ProfileViewSelectors.getPreviewSelection(getState())).toEqual(
      getSomeSelection()
    );
    dispatch(ZippedProfilesActions.returnToZipFileList());
    viewProfile('profile2.json');

    // It no longer has a selection when viewing another profile.
    expect(ProfileViewSelectors.getPreviewSelection(getState())).toEqual(
      getNoSelection()
    );
  });
});
