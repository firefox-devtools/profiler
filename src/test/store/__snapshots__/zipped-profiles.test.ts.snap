// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`reducer zipFileState will fail when trying to load an invalid profile 1`] = `
Array [
  Array [
    "UnserializationError:",
    [SyntaxError: Unexpected token 'o', "not a profile" is not valid JSON],
  ],
  Array [
    "Failed to process the profile in the archive with the following error:",
  ],
  Array [
    [Error: Unserializing the profile failed: SyntaxError: Unexpected token 'o', "not a profile" is not valid JSON],
  ],
]
`;
